### PR TITLE
TRST-M-1b: Add basket reweighting for index tokens

### PR DESCRIPTION
**Pull request by tbrent**
This PR adds automatic USD re-weighting for index RTokens (`reweightable = true`) to `setPrimeBasket()`. An additional `forceSetPrimeBasket()` is added to bypass the re-weighting, since re-weighting comes with the additional assumption that the basket is currently SOUND. 

There are tests in this PR as well as fuzz tests for `BasketLib.normalizeByPrice()` over in the `protocol-fuzzing` repo
https://github.com/reserve-protocol/protocol-fuzzing/blob/main/test/protocol/BasketLib.t.sol

Minor changes: also streamlined some naming in `requireConstantConfigTargets`. The `_` prefix for `_targetAmts` is actually a great way to indicate "this variable is different than the others it's more like scratch storage". (btw the diff is misleading in `requireConstantConfigTargets`)

### CODE REVIEW

**Comment by tbrent** [comment]
I've added an alternative flow: `forceSetPrimeBasket()` that governance can use in times of crisis to skip reweighting the basket unit to be a constant USD. Since normalizing target amounts requires a SOUND previous basket, governance may sometimes want to forcibly set the prime basket, but the default behavior should be restrictive.

**Comment by tbrent** [comment]
@pmckelvy1 @akshatmittal I'd like both your eyes on this one

File: contracts/p1/BasketHandler.sol [hunk 1]
```diff
@@ -197,21 +214,39 @@ contract BasketHandlerP1 is ComponentP1, IBasketHandler {
     //   config'.erc20s = erc20s
     //   config'.targetAmts[erc20s[i]] = targetAmts[i], for i from 0 to erc20s.length-1
     //   config'.targetNames[e] = assetRegistry.toColl(e).targetName, for e in erc20s
-    function setPrimeBasket(IERC20[] calldata erc20s, uint192[] calldata targetAmts) external {
+    function _setPrimeBasket(
+        IERC20[] calldata erc20s,
+        uint192[] memory targetAmts,
```
Comment (pmckelvy1) [inline contracts/p1/BasketHandler.sol:219]: calldata?

File: test/Main.test.ts [hunk 2]
```diff
@@ -1752,77 +1764,122 @@ describe(`MainP${IMPLEMENTATION} contract`, () => {
...
       )
     })
 
     it('Should not allow to set prime Basket with a zero amount', async () => {
+      await expect(indexBH.connect(owner).setPrimeBasket([token0.address], [0])).to.be.revertedWith(
+        'invalid target amount; must be nonzero'
+      )
+      await expect(
+        basketHandler.connect(owner).setPrimeBasket([token0.address], [0])
+      ).to.be.revertedWith('missing target weights')
```
Comment (pmckelvy1) [inline test/Main.test.ts:1877]: this revert message is different which makes me think maybe theres a different way we want to test this for non-reweightable baskets.
set token0 to the same amt but add token1 (or whatever token has same target) and set it to 0?

Comment (pmckelvy1) [reply test/Main.test.ts:1877]: also not entirely sure it matters, this test may be sufficient. it just smells weird...

File: test/Main.test.ts [hunk 3]
```diff
@@ -1833,6 +1890,12 @@ describe(`MainP${IMPLEMENTATION} contract`, () => {
           [token0.address, token1.address, token2.address, token3.address],
           [fp('0.25'), fp('0.25'), fp('0.25'), fp('0.25')]
         )
+      await basketHandler
```
Comment (pmckelvy1) [inline test/Main.test.ts:1893]: do we need to test this for indexBH as well?

File: test/Main.test.ts [hunk 4]
```diff
@@ -1855,6 +1918,14 @@ describe(`MainP${IMPLEMENTATION} contract`, () => {
       await expect(basketHandler.connect(owner).setPrimeBasket([token2.address], [fp('1')]))
         .to.emit(basketHandler, 'PrimeBasketSet')
         .withArgs([token2.address], [fp('1')], [ethers.utils.formatBytes32String('USD')])
+
+      await expect(basketHandler.connect(owner).forceSetPrimeBasket([token1.address], [fp('1')]))
```
Comment (pmckelvy1) [inline test/Main.test.ts:1922]: should we test these for indexBH as well?

File: test/Main.test.ts [hunk 5]
```diff
@@ -1926,13 +2022,24 @@ describe(`MainP${IMPLEMENTATION} contract`, () => {
...
+          .forceSetPrimeBasket([token0.address, rToken.address], [fp('0.5'), fp('0.5')])
+      ).to.be.revertedWith('invalid collateral')
+      await expect(
+        basketHandler
+          .connect(owner)
+          .forceSetPrimeBasket([token0.address, rToken.address], [fp('0.5'), fp('0.5')])
+      ).to.be.revertedWith('invalid collateral')
     })
 
     it('Should revert if target has been changed in asset registry', async () => {
```
Comment (pmckelvy1) [inline test/Main.test.ts:2045]: should this even be allowed?

**Review by pmckelvy1 — CHANGES_REQUESTED** [review]
this pr is the definition of "hold on, there is probably a simpler way to do this..."
🔥

File: contracts/p1/BasketHandler.sol (hunk 1 above)
Comment (tbrent) [reply contracts/p1/BasketHandler.sol:219]: can't be -- it gets mutated down below, conditionally

File: test/Main.test.ts (hunk 4 above)
Comment (tbrent) [reply test/Main.test.ts:1922]: yes...

File: test/Main.test.ts (hunk 5 above)
Comment (tbrent) [reply test/Main.test.ts:2045]: hard to prevent because can break `swapRegistered()` into `unregister()` and `register()`

**Review by pmckelvy1 — APPROVED** [review]
lgtm

**Review by akshatmittal — APPROVED** [review]

**Pull request merged**